homeassistant/components/sonarr/sensor.py

```diff
@@ -86,164 +97,71 @@ async def async_setup_entry(
 
     def __init__(
         self,
-        sonarr: SonarrClient,
-        host_config: PyArrHostConfiguration,
-        system_status: SystemStatus,
-        entry_id: str,
+        coordinator: SonarrDataUpdateCoordinator,
         description: SensorEntityDescription,
-        options: dict[str, Any],
     ) -> None:
         """Initialize Sonarr sensor."""
-        self.entity_description = description
-        self._attr_unique_id = f"{entry_id}_{description.key}"
-
-        self.data = {}
-        self.last_update_success = True
-        self.upcoming_days = options[CONF_UPCOMING_DAYS]
-        self.wanted_max_items = options[CONF_WANTED_MAX_ITEMS]
-
-        super().__init__(
-            sonarr=sonarr,
-            host_config=host_config,
-            system_status=system_status,
-            entry_id=entry_id,
-            device_id=entry_id,
-        )
-
-    @property
-    def available(self) -> bool:
-        """Return sensor availability."""
-        return self.last_update_success
-
-    @sonarr_exception_handler
-    async def async_update(self) -> None:
-        """Update entity."""
-        key = self.entity_description.key
-
-        if key == "diskspace":
-            self.data[key] = await self.sonarr.async_get_diskspace()
-        elif key == "commands":
-            self.data[key] = await self.sonarr.async_get_commands()
-        elif key == "queue":
-            self.data[key] = await self.sonarr.async_get_queue(
-                include_series=True, include_episode=True
-            )
-        elif key == "series":
-            self.data[key] = await self.sonarr.async_get_series()
-        elif key == "upcoming":
-            local = dt_util.start_of_local_day().replace(microsecond=0)
-            start = dt_util.as_utc(local)
-            end = start + timedelta(days=self.upcoming_days)
-
-            self.data[key] = await self.sonarr.async_get_calendar(
-                start_date=start,
-                end_date=end,
-                include_series=True,
-            )
-        elif key == "wanted":
-            self.data[key] = await self.sonarr.async_get_wanted(
-                page_size=self.wanted_max_items,
-                include_series=True,
-            )
+        super().__init__(coordinator, description)
 
     @property
     def extra_state_attributes(self) -> dict[str, str] | None:
         """Return the state attributes of the entity."""
         attrs = {}
         key = self.entity_description.key
+        data = self.coordinator.data
 
-        if key == "diskspace" and self.data.get(key) is not None:
-            for disk in self.data[key]:
+        if key == "diskspace":
+            for disk in data:
                 free = disk.freeSpace / 1024**3
                 total = disk.totalSpace / 1024**3
                 usage = free / total * 100
 
                 attrs[
                     disk.path
                 ] = f"{free:.2f}/{total:.2f}{self.unit_of_measurement} ({usage:.2f}%)"
```

Comment:
With a dedicated callable, this could also be refactored similar to `value_fn`. Maybe that's better down in a followup though. This PR is already pretty big.

Reply:
yep